pkg/cloudprovider/providers/vsphere/nodemanager.go

```diff
@@ -20,33 +20,168 @@ import (
 	"k8s.io/api/core/v1"
 	k8stypes "k8s.io/apimachinery/pkg/types"
 	"k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vclib"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"strings"
+	"errors"
+	"sync"
 )
 
 // Stores info about the kubernetes node
 type NodeInfo struct {
+	dataCenter *vclib.Datacenter
 	vm       *vclib.VirtualMachine
 	vcServer string
 }
 
 type NodeManager struct {
+	// TODO: replace map with concurrent map when k8s supports go v1.9
+
 	// Maps the VC server to VSphereInstance
 	vsphereInstanceMap map[string]*VSphereInstance
 	// Maps node name to node info.
 	nodeInfoMap map[string]*NodeInfo
+	// Maps node name to node structure
+	registeredNodes map[string]*v1.Node
+
+
+	// Mutexes
+	registeredNodesLock sync.RWMutex
+	nodeInfoLock sync.RWMutex
+}
+
+var (
+	ErrNoNodeInfo = errors.New("no node info")
+	ErrNoVSphereInstance = errors.New("no vsphere instance found")
+)
+
+func (nm *NodeManager) DiscoverNode(node *v1.Node) error {
+	for vc, vsi := range nm.vsphereInstanceMap {
+		glog.V(4).Infof("Discovering node %s in vsphere %s", node.Name, vc)
+
+		// Create context
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		err := vsi.conn.Connect(ctx)
+		if err != nil {
+			return err
+		}
+
+		var datacenterObjs []*vclib.Datacenter
+
+		glog.V(4).Infof("Finding %d datacenters in vc %s", vsi.cfg.Datacenters, vc)
+		if vsi.cfg.Datacenters == "" {
+			datacenterObjs, err = vclib.GetAllDatacenter(ctx, vsi.conn)
+			if err != nil {
+				return err
+			}
+		} else {
+			datacenters := strings.Split(vsi.cfg.Datacenters, ",")
+			for _, dc := range datacenters {
+				dc = strings.TrimSpace(dc)
+				if dc == "" {
+					continue
+				}
+				datacenterObj, err := vclib.GetDatacenter(ctx, vsi.conn, dc)
+				if err != nil {
+					return err
+				}
+				datacenterObjs = append(datacenterObjs, datacenterObj)
+			}
+		}
+		glog.V(4).Infof("Found %d datacenters in vc %s", len(datacenterObjs), vc)
+
+		type VmSearchResult struct {
+			vm *vclib.VirtualMachine
+			datacenter *vclib.Datacenter
+			err error
+		}
+
+		c1 := make(chan *VmSearchResult)
+		nodeUUID := node.Status.NodeInfo.SystemUUID
+		for _, datacenterObj := range datacenterObjs {
+			glog.V(4).Infof("Finding node %s in vc=%s and datacenter=%s", node.Name, vc, datacenterObj.Name())
+			go func(dc *vclib.Datacenter) {
+				vm, err := dc.GetVMByUUID(ctx, nodeUUID)
+				c1 <- &VmSearchResult{
+					vm:         vm,
+					err:        err,
+					datacenter: datacenterObj,
+				}
+			}(datacenterObj)
+		}
+
+		for i := 0; i < len(datacenterObjs); i++ {
+			res := <-c1
+			if res.vm != nil {
+				glog.V(4).Infof("Found node %s as vm=%+v in vc=%s and datacenter=%s",
+					node.Name, res.vm, vc, res.datacenter.Name())
+
+				nodeInfo := &NodeInfo{dataCenter: datacenterObjs[i], vm: res.vm, vcServer: vc}
+				nm.nodeInfoLock.Lock()
```

Comment:
You can move the code block that holds lock and unlocks to a separate method so that when we want to use sync.Map, then its easier to just replace the method invocation.